pyro/contrib/tracking/extended_kalman_filter.py

```diff
@@ -0,0 +1,237 @@
+        '''
+        self._mean = mean.clone()
+        self._cov = cov.clone()
+        if time is not None:
+            self._time = time
+
+        self._clear_cached()
+
+    def predict(self, dt=None, destination_time=None):
+        '''
+        Use dynamic model to predict (aka propagate aka integrate) state
+        estimate in-place.
+
+        :param dt: time to integrate over. The state time will be automatically
+                   incremented this amount unless you provide ``destination_time``.
+                   Using ``destination_time`` may be preferable for prevention of
+                   roundoff error accumulation.
+        :param destination_time: time to increment to.
+        '''
+        if dt is not None and destination_time is not None:
+            assert ((destination_time - self._time + dt) < 1e-3).all()
+        elif dt is None:
+            assert destination_time is not None
+            dt = destination_time - self._time
+        elif destination_time is None:
+            assert dt is not None
+            destination_time = self._time + dt
+
+        self._mean = self._dynamic_model(self._mean, dt)
+
+        F = self._dynamic_model.jacobian(dt)
+        Q = self._dynamic_model.process_noise_cov(dt)
+        self._cov = F.mm(self._cov).mm(F.t()) + Q
+
+        self._time = destination_time
+
+        self._clear_cached()
+
+    def innovation(self, measurement):
+        '''
+        Compute and return the innovation that a measurement would induce if
+        it were used for an update, but don't actually perform the update.
+        Assumes state and measurement are time-aligned. Useful for computing
+        Chi^2 stats and likelihoods.
+
+        :param measurement: measurement
+        :return: Innovation mean and covariance of hypothetical update.
+        :rtype: tuple(``torch.Tensor``, ``torch.Tensor``)
+        '''
+        assert self._time == measurement.time, \
+            'State time and measurement time must be aligned!'
+
+        # Compute innovation.
+        x_pv = self._dynamic_model.mean2pv(self._mean)
+        H = measurement.jacobian(x_pv)[:, :self.dimension]
+        R = measurement.cov
+        z = measurement.mean
+        z_predicted = measurement(x_pv)
+        dz = measurement.geodesic_difference(z, z_predicted)
+        S = H.mm(self._cov).mm(H.t()) + R  # innovation cov
+
+        return dz, S
+
+    def likelihood_of_update(self, measurement):
+        '''
+        Compute and return the likelihood of a potential update, but don't
+        actually perform the update. Assumes state and measurement are time-
+        aligned. Useful for gating and calculating costs in assignment problems
+        for data association.
+
+        :param: measurement.
+        :return: Likelihood of hypothetical update.
+        '''
+        dz, S = self.innovation(measurement)
+        return torch.exp(dist.MultivariateNormal(torch.zeros_like(S), S)
```

Comment:
This is suspicious: the mean and cov of a `MultivariateNormal` have different shape Did you instead intend
```py
dist.MultivariateNormal(S.new_zeros(S.shape[-1]), S)
```

Reply:
wow i wonder why that didnt throw a shape error somewhere